Aromero/spiral provider

Ossyahn opened the pull request:
This algorithm walks the given grid in a spiral pattern from the center outwards and saves all the available tiles in a list respecting the spiral-shaped order. After all available tiles are saved in the list (finishing when reaching a max distance from the center of the spiral), the list is used to choose which tiles to return in a tiles array (the amount of which is given by parameter). 

For every tile to put in the array the algorithm walks the ordered list of available tiles from the beginning, and at every step decides by a probability (30%) to return that tile or not, and if it's returned removes that tile from the ordered list so the returned array have unique tiles.

That gives an array of tiles clustered around the center of the spiral, as the tiles are more prone to be chosen the closer to the center of the spiral they are.

Ossyahn commented on Assets/ScriptableObjects/Grid/TestGridData.asset:
```diff
@@ -0,0 +1,16 @@
+%YAML 1.1
```
Removed

Ossyahn commented on Assets/Scripts/Grid/IGrid.cs:
```diff
@@ -36,7 +36,10 @@ public interface IGrid {
+        /// <summary>
+        /// The bounds of the rectangle containing this grid, in tile coordinates.
+        /// </summary>
+        Rect TileBounds { get; }
```
Moved to helper method as in #5

Ossyahn commented on Assets/Scripts/Grid/Positioning/SpiralSequenceRandomPositionProvider.cs:
```diff
@@ -0,0 +1,119 @@
+    {
+        private IRandomProvider _randomProvider;
+        private IGridPositionCalculator _gridPositionCalculator;
+        private const float TileChooseProbability = .3f;
```
Constant names changed to convention

Ossyahn commented on Assets/Scripts/Grid/Positioning/SpiralSequenceRandomPositionProvider.cs:
```diff
@@ -0,0 +1,119 @@
+                
+                //If at this point the position is (center.x - maxDistance, center.y - maxDistance) the spiral is
+                //complete for maxDistance.
+                if (currentPosition == spiralEndTile) break;
```
one line if not one line anymore

Ossyahn commented on Assets/Scripts/Grid/Positioning/SpiralSequenceRandomPositionProvider.cs:
```diff
@@ -0,0 +1,119 @@
+                k++;
+            } while (true);
+
+            return ChooseRandomPositionsFromAvailableTiles(availableTiles, numTilesToGenerate);
```
Thanks!

Ossyahn commented on Assets/Scripts/Grid/Positioning/SpiralSequenceRandomPositionProvider.cs:
```diff
@@ -0,0 +1,119 @@
+        }
+
+        private void AddIfAvailableTile(LinkedList<Vector2> availableTiles, IGrid grid, Vector2 tile) {
+            if (grid.TileBounds.Contains(tile)) {
```
Already moved to a helper method

Ossyahn commented on Assets/Scripts/Grid/Positioning/SpiralSequenceRandomPositionProvider.cs:
```diff
@@ -0,0 +1,119 @@
+using Vector2 = UnityEngine.Vector2;
+
+namespace Grid.Positioning
+{
```
remaining curly braces changed

Ossyahn commented on Assets/Scripts/Grid/Positioning/SpiralSequenceRandomPositionProvider.cs:
```diff
@@ -0,0 +1,119 @@
+                    if (randomNumber <= TileChooseProbability) {
+                        tiles[chosenTiles] = tile;
+                        chosenTiles++;
+                        availableTiles.Remove(tile);
```
I suppose that the time complexity of removing an element is O(L), where L is the list size, as it has to do a linear search.

`ChooseRandomPositionsFromAvailableTiles` would be O(L<sup>2</sup>). Well, it usually would be O(L<sup>2</sup>*M) where M is MaxTries, but because in this case is a known constant, which is 100, wouldn't be considered.

But considering how this function works clustering the chosen tiles near the center of the spiral (at the beginning of the list) and because there's a break every time the tile is chosen, there would be a bunch of amortized time.

Ossyahn commented:
The required changes have been implemented

dcasadevall commented on .gitignore:
```diff
@@ -34,3 +34,6 @@ sysinfo.txt
 *.apk
 *.unitypackage
+.idea/
+Assets/ScriptableObjects/Grid/TestGridData.asset.meta
```
These two files should not be in .gitignore.
Normally you put files that you will be creating (and only you) in your global gitignore list (right click in sourcetree file and click "ignore...")

Ossyahn replied:
That's exactly what I did, but I will check to make sure that those changes go to my global .gitignore instead of the one in the project.

dcasadevall commented on Assets/Scripts/Grid/Grid.cs:
```diff
@@ -40,5 +40,6 @@ public class Grid : IGrid {
             _numTilesX = System.Math.Max(1, gridData.NumTilesX);
             _numTilesY = System.Math.Max(1, gridData.NumTilesY);
         }
+        
```
Please don't commit this extra space.

Ossyahn replied:
Yeah, this was with the diff tool, something happened. I will be more careful next time.

dcasadevall commented on Assets/Scripts/Grid/IGrid.cs:
```diff
@@ -50,5 +49,12 @@ public static class GridExtensions {
+            return new Rect(Vector2.zero,
+                            new Vector2(grid.NumTilesX, grid.NumTilesY));
+    	}
+	}
```
There are some indentation issues here.

Ossyahn replied:
Same here with the diff tool.

dcasadevall commented on Assets/Scripts/Grid/Positioning/SpiralSequenceRandomPositionProvider.cs:
```diff
@@ -0,0 +1,109 @@
+using Math.Random;
+using UnityEngine;
+
+namespace Grid.Positioning
```
The curly bracer is still on the next line <,<

dcasadevall commented on Assets/Scripts/Grid/Positioning/SpiralSequenceRandomPositionProvider.cs:
```diff
@@ -0,0 +1,109 @@
+    /// It returns a list of random unique positions walking through the grid from the center outwards in a
+    /// spiral pattern, deciding at every step to pick that position or not by a probability range (30%) 
+    /// </summary>
+    public class SpiralSequenceRandomPositionProvider : IRandomGridPositionProvider{
```
add a space before the curly brace:

`IRandomGridPositionProvider {`

Ossyahn replied:
Almost

dcasadevall commented on Assets/Scripts/Grid/Positioning/SpiralSequenceRandomPositionProvider.cs:
```diff
@@ -0,0 +1,109 @@
+        private IGridPositionCalculator _gridPositionCalculator;
+
+        public SpiralSequenceRandomPositionProvider(IGridPositionCalculator gridPositionCalculator, 
+                                                    IRandomProvider randomProvider){
```
space before curly brace

dcasadevall commented on Assets/Scripts/Grid/Positioning/SpiralSequenceRandomPositionProvider.cs:
```diff
@@ -0,0 +1,109 @@
+            _gridPositionCalculator = gridPositionCalculator;
+        }
+        
+        public Vector2[] GetRandomUniquePositions(IGrid grid, int maxDistanceFromCenter, int numTilesToGenerate){
```
space before curly brace

dcasadevall commented on Assets/Scripts/Grid/Positioning/SpiralSequenceRandomPositionProvider.cs:
```diff
@@ -0,0 +1,109 @@
+            return ChooseRandomPositionsFromAvailableTiles(availableTiles, numTilesToGenerate);
+        }
+
+        private void AddIfAvailableTile(LinkedList<Vector2> availableTiles, IGrid grid, Vector2 tile) {
```
A better name for this method would be `AddTileIfInsideGridBounds`

Ossyahn replied:
I agree.

Ossyahn commented:
Ok, I think that's it, finally. 

About the braces, I realized the change in style midways and changed it in Rider preferences, and I left out a bunch, that were changed manually (thinking about it, I'm sure Rider has an option to update a file to fit the preferences...). Please, don't hesitate in pointing more style issues.

For example I realized that the Rider default of writing properties with just get or set in a single line is not the one that you use, and I've changed it as well in preferences. Correct me if that way is not how you like it.

dcasadevall commented on Assets/Scripts/Grid/IGrid.cs:
```diff
@@ -32,7 +32,6 @@ public interface IGrid {
         /// TileSize will always be >=1.
         /// </summary>
         uint TileSize { get; }
-        
```
Please don't remove this line

dcasadevall replied:
Actually I see why you did remove it.
I guess it's fine with the comments below.
Let's merge this!
